src/lib.rs

```diff
@@ -2,269 +2,225 @@
 //
 // SPDX-License-Identifier: MIT OR Apache-2.0
 
-/*! The library provide tools for handling compressed and archive files
-
-# Examples
-```no_run
-use compress_tools;
-
-let dir = tempfile::tempdir().unwrap();
-compress_tools::uncompress("tests/fixtures/tree.tar.gz", dir.path(), compress_tools::Kind::TarGZip).unwrap();
-```
-*/
-
-mod uncompress;
-
-pub use uncompress::{Error, Result};
-
-use std::path::Path;
-
-/// Type of compressed file or archive
-#[derive(Copy, Clone, Debug)]
-pub enum Kind {
-    TarGZip,
-    TarBZip2,
-    TarXz,
-    TarLZMA,
-    TarLZip,
-    Tar,
-    Zip,
-
-    GZip,
-    BZip2,
-    Xz,
-    LZip,
-    LZMA,
+pub mod error;
+pub mod ffi;
+
+pub use error as compress;
+use error::*;
+use ffi::*;
+use std::{
+    convert::TryInto,
+    ffi::{CStr, CString},
+    io::Read,
+    os::{raw::c_void, unix::io::AsRawFd},
+    path::Path,
+    slice,
+};
+
+const BUFFER_SIZE: usize = 10240;
+
+struct Pipe {
+    reader: Box<dyn Read>,
+    buffer: Vec<u8>,
 }
 
-/// Uncompress a archive of known [Kind](Kind) pointed by source to target
-/// location for file
-pub fn uncompress<P1, P2>(source: P1, target: P2, kind: Kind) -> Result<()>
+/// Reads the input buffer and extract to another buffer.
+///
+/// # Arguments
+///
+/// * `source` - Reader.
+/// * `target` - Writer.
+///
+/// # Example
+///
+/// uncompress_file(std::io::Read, std::io::Write)?;
+pub fn uncompress_file<R: 'static, W>(source: R, target: W) -> compress::Result<()>
 where
-    P1: AsRef<Path>,
-    P2: AsRef<Path>,
+    R: std::io::Read,
```

Comment:
Since we have `R`'s type definition on `where` clause, I think it's better to move the `'static` lifetime requirement to the where clause aswell.